### PR TITLE
Support files with type comment syntax errors

### LINKED ISSUE #3593: Let black run on files even if `ast.parse` with `type_comments=True` fails
**Is your feature request related to a problem? Please describe.**

Right now, if a file parsing because of invalid type comments (even if the code runs just fine), black won't format the file.

For example:

```python
import typing

def foo(
    # type: typing.List
    bar,
):
    print(bar)

foo(10)
```

While the code still runs:

```console
$ python3 asd.py            
10
```

Black no longer formats it:

```console
$ black asd.py        
error: cannot format asd.py: cannot use --safe with this file; failed to parse source file AST: invalid syntax (<unknown>, line 4)
This could be caused by running Black with an older Python version that does not support new syntax used in your source file.

Oh no! 💥 💔 💥
1 file failed to reformat.
```

This is because `ast.parse` on that code fails if you pass `type_comments=True`:

```pycon
>>> import ast
>>> ast.parse(open('asd.py').read(), type_comments=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 4
    # type: typing.List
            ^^^^^^^^^^^
SyntaxError: invalid syntax
```

(Black does work if you do `black asd.py --fast`, as it no longer does the AST based check.)

**Describe the solution you'd like**

If `ast.parse(..., type_comments=True)` fails for the before/after check, maybe doing it with `type_comments=False` is fine, just for that file?

**Describe alternatives you've considered**

None so far.

**Additional context**

I don't think this can be considered an upstream bug, as this is indeed invalid syntax for a type checker's perspective. But black shouldn't care if the types are correct, in any way shape or form.

### PR DESCRIPTION
**Pull request by tusharsadhwani**
### Description

Resolves #3593 

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

File: tests/test_black.py [hunk 1]
```diff
@@ -1938,6 +1938,14 @@ def test_equivalency_ast_parse_failure_includes_error(self) -> None:
         err.match("invalid character")
         err.match(r"\(<unknown>, line 1\)")
 
+    def test_type_comment_syntax_error(self) -> None:
```
Comment (JelleZijlstra) [inline tests/test_black.py:1941]: Would prefer a test in test_format.py where we don't have to write a new test function.

File: src/black/parsing.py [hunk 2]
```diff
@@ -175,11 +180,17 @@ def parse_ast(src: str) -> Union[ast.AST, ast3.AST]:
     first_error = ""
     for version in sorted(versions, reverse=True):
         try:
-            return parse_single_version(src, version)
+            return parse_single_version(src, version, type_comments=True)
         except SyntaxError as e:
             if not first_error:
                 first_error = str(e)
 
+            # Try to parse without type comments
```
Comment (JelleZijlstra) [inline src/black/parsing.py:188]: Need to think more about whether we should have a separate loop instead. I *think* that should never affect behavior, just performance.

File: src/black/parsing.py [hunk 3]
```diff
@@ -148,24 +148,29 @@ def lib2to3_unparse(node: Node) -> str:
 
 
 def parse_single_version(
-    src: str, version: Tuple[int, int]
+    src: str, version: Tuple[int, int], type_comments: bool
```
Comment (JelleZijlstra) [inline src/black/parsing.py:151]:
```suggestion
    src: str, version: Tuple[int, int], *, type_comments: bool
```
Let's make it keyword-only

File: src/black/parsing.py (hunk 2 above)
Comment (tusharsadhwani) [reply src/black/parsing.py:188]: I'll move it to an outer for loop anyway. Just for guaranteed backwards compatibility.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by tusharsadhwani** [comment]
@ichard26 can you help with re-triggering the workflows?

**Comment by JelleZijlstra** [comment]
@ichard26 do you still want to look at this or can I merge it?

**Review by ichard26 — APPROVED** [review]
I wanted to take a look because this PR falls back to using the built-in ast module even on Python 3.7 which we avoided doing previously, but that's a really minor edge case — so it's fine. All it might do is make some files with invalid type comments still unparsable.

**Comment by ichard26** [comment]
Thanks for the PR @tusharsadhwani!

**Pull request merged**